raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,298 @@
+
+        // Call updateTransfer in the corresponding TokenNetwork
+        token_network.updateNonClosingBalanceProof(
+            closing_participant,
+            non_closing_participant,
+            balance_hash,
+            nonce,
+            additional_hash,
+            closing_signature,
+            non_closing_signature
+        );
+
+        emit NewBalanceProofReceived(
+            reward_amount,
+            nonce,
+            msg.sender,
+            non_closing_participant
+        );
+    }
+
+    /// @notice Called after a monitored channel is settled in order for MS to claim the reward
+    /// Can be called once per settled channel by everyone on behalf of MS
+    /// @param token_network_address Address of the Token Network in which the channel
+    /// @param closing_participant Address of the participant of the channel that called close
+    /// @param non_closing_participant The other participant of the channel
+    function claimReward(
+        address token_network_address,
+        address closing_participant,
+        address non_closing_participant
+    )
+        public
+        returns (bool)
+    {
+        TokenNetwork token_network = TokenNetwork(token_network_address);
+        bytes32 channel_identifier = token_network.getChannelIdentifier(
+            closing_participant,
+            non_closing_participant
+        );
+        bytes32 reward_identifier = keccak256(channel_identifier, token_network_address);
+
+        // Only allowed to claim, if channel is settled
+        // Channel is settled if it's data has been deleted
+        uint8 channel_state;
+        (, , channel_state) = token_network.getChannelInfo(
+            closing_participant,
+            non_closing_participant
+        );
+        // If channel.state is zero it means it's either non-existing or settled
+        require(channel_state == 0);
+
+        Reward storage reward = rewards[reward_identifier];
+
+        // Make sure that the Reward exists
+        require(reward.reward_sender_address != 0x0);
+
+        // Deduct reward from raiden_node deposit
+        balances[reward.reward_sender_address] -= reward.reward_amount;
+        // Add reward to the monitoring services' balance.
+        // This minimizes the amount of gas cost
+        // Only use token.transfer in the withdraw function
+        balances[reward.monitoring_service_address] += reward.reward_amount;
+
+        emit RewardClaimed(
+            reward.monitoring_service_address,
+            reward.reward_amount,
+            reward_identifier
+        );
+
+        // delete storage
+        delete rewards[reward_identifier];
+    }
+
+    /// @notice Withdraw depositted tokens.
+    /// Can be called by addresses with a deposit as long as they have a positive balance
+    /// @param amount Amount of tokens to be withdrawn
+    function withdraw(uint amount) public {
+        require(balances[msg.sender] - amount >= 0);
```

Comment:
This does not protect from underflows, we need a `require(balances[msg.sender] >= amount);`